src/meta/src/hummock/compaction_schedule_policy.rs

```diff
@@ -0,0 +1,679 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::{BTreeMap, HashMap, HashSet};
+use std::sync::Arc;
+
+use rand::Rng;
+use risingwave_hummock_sdk::HummockContextId;
+use risingwave_pb::hummock::compact_task::TaskStatus;
+use risingwave_pb::hummock::{CompactTask, CompactTaskAssignment, SubscribeCompactTasksResponse};
+use tokio::sync::mpsc::Receiver;
+
+use super::Compactor;
+use crate::MetaResult;
+
+const STREAM_BUFFER_SIZE: usize = 4;
+
+/// The implementation of compaction task scheduling policy.
+pub trait CompactionSchedulePolicy: Send + Sync {
+    /// Get next idle compactor to assign task.
+    fn next_idle_compactor(
+        &mut self,
+        compactor_assigned_task_num: &HashMap<HummockContextId, u64>,
+        compact_task: Option<&CompactTask>,
+    ) -> Option<Arc<Compactor>>;
+
+    /// Get next compactor to assign task.
+    fn next_compactor(&mut self, compact_task: Option<&CompactTask>) -> Option<Arc<Compactor>>;
+
+    fn random_compactor(&mut self, compact_task: Option<&CompactTask>) -> Option<Arc<Compactor>>;
```

Comment:
So the interface should look like this:

```rust
trait CompactionSchedulePolicy {
    fn next_idle_compactor(&self, ...) -> Option<Arc<Compactor>>;
    fn assign_compaction_task(&mut self, context_id: HummockContextId, task: &CompactTask);
}
```

Putting the compactor picking and policy state update into `assign_compaction_task`, the logic will be:

- Acquire compaction lock.
- Pick compactor from `CompactorManager`.
- Assign the task to `HummockManager`.
- Update policy state. This step should not produce any error.

Reply:
BTW, we don't cancel compaction task when `assign_compaction_task` encounters an error. Is this by design?\

```rust
match self
                .hummock_manager
                .assign_compaction_task(&compact_task, compactor.context_id())
                .await
            {
                Ok(_) => {
                    tracing::trace!(
                        "Assigned compaction task. {}",
                        compact_task_to_string(&compact_task)
                    );
                }
                Err(err) => {
                    tracing::warn!(
                        "Failed to assign compaction task to compactor {}: {:#?}",
                        compactor.context_id(),
                        err
                    );
                    match err {
                        Error::InvalidContext(_) | Error::CompactorUnreachable(_) => {
                            self.compactor_manager
                                .remove_compactor(compactor.context_id());
                        }
                        _ => {}
                    }
                    continue 'send_task;
                }
            }
```

Reply:
> BTW, we don't cancel compaction task when assign_compaction_task encounters an error. Is this by design?

This looks like a bug. cc @zwang28 Can you help take a look?

Reply:
> The only restriction will be that we're no longer able to specify a compactor to assign task, but that's probably OK.

If we really need that, it is still doable with another policy (e.g. ManualSchedulePolicy).

Reply:
> This looks like a bug. cc @zwang28 Can you help take a look?

After his recent refactor (#5240 ), I think it is fixed now.

Reply:
> BTW, we don't cancel compaction task when assign_compaction_task encounters an error. Is this by design?

Yes it was by design before https://github.com/risingwavelabs/risingwave/pull/5240,  where we rely on `continue 'send_task;` to retry assignment rather than cancelling task.

https://github.com/risingwavelabs/risingwave/pull/5240 unifies the cancellation path.